### PR TITLE
Thea Pallet: Set Rotation

### LINKED ISSUE #645: Add `From` field for `ApprovedWithdraw` Struct

### LINKED ISSUE #647: Pallet Side Check to not allow `Withdrawals` during Key Rotation

### PR DESCRIPTION
**Pull request by Gauthamastro**
Relayers should use Queued Relayers of Thea Staking Pallet and Active Relayers of Thea pallet for operations. Thea pallet will generate an ingress message when the key updation is complete on foreign chain after which new relayers can start processing withdrawals.  Queued Relayers of Thea Staking Pallet is used to pre-generate Thea Public key and Offline Stages  for next set and update the on-chain state about it.

### CODE REVIEW

**Comment by Gauthamastro** [comment]
TODO: Sending new public key back to the on-chain state

File: pallets/thea/src/lib.rs [hunk 1]
```diff
@@ -481,6 +510,42 @@ pub mod pallet {
...
+			// Call host function with current_active_relayer_set, signature, bit_map, verify nonce
+			ensure!(
+				thea_primitives::thea_ext::bls_verify(
+					&bls_signature,
+					bit_map,
+					&tx_hash.encode(),
+					&current_active_relayer_set
+				),
+				Error::<T>::BLSSignatureVerificationFailed
+			);
```
Comment (felixfaisal) [inline pallets/thea/src/lib.rs:852]: We should track this hash, as there exist a scenario that can lead to replay attack if relayers remain unchanged over different sets .

**Review by felixfaisal — CHANGES_REQUESTED** [review]

**Comment by felixfaisal** [comment]
Resolves #647 
Resolves #645 (In events rather than in `struct` as mentioned in the issue)

File: pallets/thea/src/lib.rs [hunk 2]
```diff
@@ -468,23 +514,183 @@ pub mod pallet {
...
+				// as the new QQPublicKey rather than returning an Error
+				<QueuedQueuedTheaPublicKey<T>>::insert(network, public_key);
 			}
+
+			// Add the new one to queued_queued
+			Ok(())
+		}
+
+		#[pallet::weight(1000)]
+		pub fn thea_relayers_reset_rotation(
```
Comment (serhii-temchenko) [inline pallets/thea/src/lib.rs:680]: Please add docblock as on other extrinsics

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/thea/src/lib.rs (hunk 2 above)
Comment (felixfaisal) [reply pallets/thea/src/lib.rs:680]: Done ✅

**Review by serhii-temchenko — APPROVED** [review]

**Review by felixfaisal — APPROVED** [review]

**Pull request merged**